lib/ntlmauth/ntlmauth.cc

```diff
@@ -108,10 +108,19 @@ ntlm_fetch_string(const ntlmhdr *packet, const int32_t packet_size, const strhdr
     int32_t o = le32toh(str->offset);
     // debug("ntlm_fetch_string(plength=%d,l=%d,o=%d)\n",packet_size,l,o);
 
-    if (l < 0 || l > NTLM_MAX_FIELD_LENGTH || o + l > packet_size || o == 0) {
-        debug("ntlm_fetch_string: insane data (pkt-sz: %d, fetch len: %d, offset: %d)\n", packet_size,l,o);
+    if (l < 0 || NTLM_MAX_FIELD_LENGTH < l) {
+        debug("ntlm_fetch_string: insane string length (pkt-sz: %d, fetch len: %d, offset: %d)\n", packet_size,l,o);
         return rv;
     }
+    else if (o <= 0 || packet_size < o) {
+        debug("ntlm_fetch_string: insane string offset (pkt-sz: %d, fetch len: %d, offset: %d)\n", packet_size,l,o);
+        return rv;
+    }
+    else if (l > packet_size - o) {
+        debug("ntlm_fetch_string: truncated string data (pkt-sz: %d, fetch len: %d, offset: %d)\n", packet_size,l,o);
+        return rv;
+    }
+
     rv.str = (char *)packet + o;
     rv.l = 0;
     if ((flags & NTLM_NEGOTIATE_ASCII) == 0) {
```

Comment:
The loop below appears to assume that `l` is even when dealing with `NTLM_NEGOTIATE_ASCII` but we do not actually check for that.

Reply:
Huh?  the loop iterator ```len``` is set to ```l>>1``` which erases the final bit. This is looping the number of UTF-16 characters, not number of octets.

Reply:
You may have been fooled by the check that ASCII flag is set to **false** (```== 0```).

Reply:
> the loop iterator `len` is set to `l>>1` which erases the final bit. This is looping the number of whole UTF-16 characters, not number of octets.

Exactly. When `l` is 3 bytes, and we expect to see 2-byte characters, then something went wrong, and we should probably reject (or at least warn about) input instead of ignoring the fact that the total size of parsed 2-byte characters does not match the specified total length. In other words, we should reject or warn about input containing half-characters.